appserver/tests/test-micro-jar/src/test/java/fish/payara/admin/cli/AsAdminIntegrationTest.java

```diff
@@ -0,0 +1,198 @@
+ * General Public License Version 2 only ("GPL") or the Common Development
+ * and Distribution License("CDDL") (collectively, the "License").  You
+ * may not use this file except in compliance with the License.  You can
+ * obtain a copy of the License at
+ * https://github.com/payara/Payara/blob/master/LICENSE.txt
+ * See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing the software, include this License Header Notice in each
+ * file and include the License file at glassfish/legal/LICENSE.txt.
+ *
+ * GPL Classpath Exception:
+ * The Payara Foundation designates this particular file as subject to the "Classpath"
+ * exception as provided by the Payara Foundation in the GPL Version 2 section of the License
+ * file that accompanied this code.
+ *
+ * Modifications:
+ * If applicable, add the following below the License Header, with the fields
+ * enclosed by brackets [] replaced by your own identifying information:
+ * "Portions Copyright [year] [name of copyright owner]"
+ *
+ * Contributor(s):
+ * If you wish your version of this file to be governed by only the CDDL or
+ * only the GPL Version 2, indicate your decision by adding "[Contributor]
+ * elects to include this software in this distribution under the [CDDL or GPL
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package fish.payara.admin.cli;
+
+import static fish.payara.micro.ClusterCommandResult.ExitStatus.FAILURE;
+import static fish.payara.micro.ClusterCommandResult.ExitStatus.SUCCESS;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+import org.hamcrest.CoreMatchers;
+import org.junit.After;
+import org.junit.Before;
+
+import fish.payara.micro.ClusterCommandResult;
+import fish.payara.micro.ClusterCommandResult.ExitStatus;
+import fish.payara.test.util.PayaraMicroServer;
+
+public abstract class AsAdminIntegrationTest {
+
+    private static final String UNSATISFIED_DEPENDENCY_EXCEPTION_CLASS_NAME = 
+            "org.jvnet.hk2.config.UnsatisfiedDependencyException";
+    private static final String UNACCEPTABLE_VALUE_EXCEPTION_CLASS_NAME = 
+            "org.glassfish.common.util.admin.UnacceptableValueException";
+
+    protected final PayaraMicroServer server = PayaraMicroServer.DEFAULT;
+
+    @Before
+    public void serverSetUp() throws Exception {
```

Comment:
Could this potentially be a before and after class to optimise the test running?

Reply:
I started this way but then decided that the more flexible reuse is to encapsulate the server handling in a class on its own so it became the `PayaraMicroServer` class. The extra class allows to create multiple servers but there is the `PayaraMicroServer.DEFAULT` which can be used to share the same server for all the tests. It is special in the way that calling `stop()` will not stop the default server which instead uses a JVM shutdown hook. I documented this in the `PayaraMicroServer` class. The base class still does the `start()` and `stop()` in before and after to be independent of the server instance used just in case that should change. The idea was that we can pretend the default instance is just as all other but internally all tests using the default will share the same server and thereby only pay overhead of starting/stopping once.